modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/internal/upgrade/v3_1_4/test/SXPBlueprintAndSXPElementUpgradeProcessTest.java

```diff
@@ -0,0 +1,315 @@
+			_serviceContext);
+	}
+
+	private void _assertElementUpgraded(String externalReferenceCode)
+		throws Exception {
+
+		SXPElement sxpElement =
+			_sxpElementLocalService.fetchSXPElementByExternalReferenceCode(
+				externalReferenceCode, TestPropsValues.getCompanyId());
+
+		JSONAssert.assertEquals(
+			sxpElement.getElementDefinitionJSON(),
+			_getExpectedElementDefinitionJSON(externalReferenceCode),
+			JSONCompareMode.STRICT);
+	}
+
+	private void _assertSXPBlueprint(long sxpBlueprintId) throws Exception {
+		SXPBlueprint sxpBlueprint = _sxpBlueprintLocalService.fetchSXPBlueprint(
+			sxpBlueprintId);
+
+		Assert.assertNotNull(sxpBlueprint);
+
+		JSONAssert.assertEquals(
+			_getExpectedInstancesJSON(), sxpBlueprint.getElementInstancesJSON(),
+			JSONCompareMode.STRICT);
+	}
+
+	private String _createAssetCategoryExternalReferenceCodeLabel(
+		AssetCategory assetCategory) {
+
+		return StringBundler.concat(
+			assetCategory.getName(), " (ERC: ",
+			assetCategory.getExternalReferenceCode(), ")");
+	}
+
+	private String _createAssetCategoryIDLabel(AssetCategory assetCategory) {
+		return StringBundler.concat(
+			assetCategory.getName(), " (ID: ", assetCategory.getCategoryId(),
+			")");
+	}
+
+	private void _createOldElement(String externalReferenceCode)
+		throws Exception {
+
+		SXPElement sxpElement =
+			_sxpElementLocalService.fetchSXPElementByExternalReferenceCode(
+				externalReferenceCode, TestPropsValues.getCompanyId());
+
+		if (sxpElement != null) {
+			sxpElement.setElementDefinitionJSON(RandomTestUtil.randomString());
+
+			_sxpElementLocalService.updateSXPElement(sxpElement);
+		}
+		else {
+			_sxpElementLocalService.addSXPElement(
+				externalReferenceCode, TestPropsValues.getUserId(),
+				Collections.singletonMap(LocaleUtil.US, StringPool.BLANK),
+				RandomTestUtil.randomString(), StringPool.BLANK,
+				StringPool.BLANK, true, StringPool.BLANK,
+				Collections.singletonMap(
+					LocaleUtil.US, RandomTestUtil.randomString()),
+				0,
+				ServiceContextTestUtil.getServiceContext(
+					TestPropsValues.getCompanyId(),
+					TestPropsValues.getGroupId(), TestPropsValues.getUserId()));
+		}
+	}
+
+	private String _getExpectedElementDefinitionJSON(
+		String externalReferenceCode) {
+
+		Class<?> clazz = getClass();
+
+		return StringUtil.read(
+			clazz,
+			StringBundler.concat(
+				"dependencies/", clazz.getSimpleName(), StringPool.PERIOD,
+				testName.getMethodName(), StringPool.PERIOD,
+				StringUtil.toLowerCase(externalReferenceCode), ".json"));
+	}
+
+	private String _getExpectedInstancesJSON() {
+		Class<?> clazz = getClass();
+
+		String elementInstancesJSON = StringUtil.read(
+			clazz,
+			StringBundler.concat(
+				"dependencies/", clazz.getSimpleName(), StringPool.PERIOD,
+				testName.getMethodName(), ".after.json"));
+
+		elementInstancesJSON = StringUtil.replace(
+			elementInstancesJSON, "$ASSET_CATEGORY_EXTERNAL_REFERENCE_CODE_1$",
+			_globalGroup.getExternalReferenceCode() + "&&" +
```

Comment:
You did a method for `_createAssetCategoryExternalReferenceCodeLabel` but not for `_globalGroup.getExternalReferenceCode() + "&&" +
				_assetCategory2.getExternalReferenceCode()` Please extract this too it will look better.

Reply:
Please rename the method for a proper one I used a generic name